### ISSUE TITLE
ssh/tailssh: Cannot login as LDAP user on Linux host

### ISSUE THREAD

**Issue by DilumAluthge** [tailscale/tailscale#4945]
### What is the issue?

I've set up Tailscale SSH on a Linux server where most of the users are from LDAP. For example:

```
# cat /etc/passwd | grep jsmith
# cat /etc/shadow | grep jsmith
# getent passwd | grep jsmith
jsmith:x:123456789:123456789:Smith, John:/home/jsmith:/bin/bash
```

When I try to SSH in as an LDAP user, I am denied with the following message:

```
$ ssh jsmith@myserver
failed to lookup jsmith
jsmith@myserver: Permission denied (tailscale).
```

I'm guessing that this corresponds to an error being returned by this call to `user.Lookup`:

https://github.com/tailscale/tailscale/blob/1d04e01d1e62e0f83a24c0e2963c1b74db33c00d/ssh/tailssh/tailssh.go#L208-L215

If I try to SSH in as a local (non-LDAP) user, it works successfully.

### Steps to reproduce

To reproduce this issue, first make sure that `LDAPUSER` is an LDAP user on your server. Then run:

```
ssh LDAPUSER@myserver
```

### Are there any recent changes that introduced the issue?

I recently did the following:
1. Updated Tailscale (by doing `sudo yum update tailscale`).
2. Ran `sudo tailscale up --ssh` to start the Tailscale SSH server.

### OS

Linux

### OS version

CentOS Linux release 7.9.2009 (Core)

### Tailscale version

1.26.1

### Bug report

BUG-53b053b3be60de1d1ace2edd177cb07c87b1f3b1a4c49b3b307d74b854fe7801-20220627044844Z-95747bf68628c167

### Other information

I installed Tailscale by doing adding the Tailscale repository and then doing `sudo yum install tailscale`.

```
# tailscale --version
1.26.1
  tailscale commit: b3687b70758d29f5fd6021c34d58eba7381812c0
  other commit: bcd70ce328082f464f89735d9adfa5bcd90ccf7e
  go version: go1.18.3-ts04d67b90d8
```

```
# cat /etc/system-release
CentOS Linux release 7.9.2009 (Core)
# uname -a
Linux myserver 3.10.0-1160.59.1.el7.x86_64 #1 SMP Wed Feb 23 16:47:03 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by bradfitz** [comment]
Ah, probably because the binaries we distribute aren't cgo binaries, so we're not using the `lookupUser` in `os/user` that uses `C.getpwnam_r`, etc, and are thus only using `os/user`.

If you build `tailscaled` yourself it'll probably work. (which isn't a very satisfying answer in general, but if you're curious to try)

**Comment by DilumAluthge** [comment]
I can confirm that everything works correctly if I build `tailscaled` myself!

Of course, as you said, it would be nice to have an easier solution for the long term.

**Comment by porkcharsui** [comment]
This issue also impacts users populated by [nsscache](https://github.com/google/nsscache).

The suggestion to build `tailscaled` also unblocked usage for this case.

**Comment by DentonGentry** [comment]
Current status of this issue:
- Compiling `tailscaled` locally links in libc and uses the C implementation of lookupUser, which knows about NSS and PAM and so on.
- Linking to libc isn't a solution we can use more broadly: we'd turn a relatively narrow issue in SSH into a widespread failure of tailscaled to start on systems where we got the choice of glibc/musl/uclibc wrong or where glibc versioning breaks it.

Instead, https://github.com/tailscale/tailscale/pull/8098 uses `getent` to get information about users. `getent` is supplied by the system and has to support LDAP or NIS+ or what have you.

This support is included in 1.41.48 and later in the unstable (development) track: https://pkgs.tailscale.com/unstable/, and will be in the 1.42.0 stable release.

We don't have access to many variations of LDAP, nsscache, and so on. If anyone who previously compiled `tailscaled` locally is willing to try 1.41.48, please let us know if it works and describe a bit about the auth configuration.

**Comment by DentonGentry** [comment]
We have one report from a site using LDAP users that 1.41.48 works for SSH for them.

**Comment by JordanSekky** [comment]
I can confirm that 1.41.48 works for ssh me as well.

Ubuntu 22.04 using sssd to draw users from a Synology LDAP server.

**Comment by SoulEater45** [comment]
I just installed 1.40.1 on our servers running auth with LDAP and so far there are no problems!
